File: LeetCodeProblems/Problems/LeetCode89GrayCode.cs
namespace LeetCodeProblems.Problems;

public class LeetCode89GrayCode
{
    public IList<int> GrayCode(int n)
    {
        int length = 1 << n;
        var result = new List<int>(length);
        for (int i = 0; i < length; i++)
        {
            result.Add(i ^ (i >> 1));
        }

        return result;
    }
}

File: LeetCodeProblemsCMake/Problems/LeetCode89GrayCode.cpp
//
// Created by saeed on 1/23/25.
//

#include "LeetCode89GrayCode.h"

#include <complex>
#include <unordered_map>
#include <unordered_set>

#include "../Auxilary/Auxilaries.h"

std::vector<int> LeetCode89GrayCode::grayCode(int n) {
    const int count = 1 << n;
    std::vector result(count, 0);
    for (int i = 0; i < count; i++) {
        result[i] = i ^ i >> 1;
    }
    return result;
}

void LeetCode89GrayCode::Run() {
    int n = 4;
    auto result = grayCode(n);
    Auxilaries::PrintVector(result);
}
